src/database-sqlite.ts

```diff
@@ -30,12 +29,39 @@ export function initStorage (sqlitedb: string): Promise<void> {
   db.execute('CREATE TABLE IF NOT EXISTS Data(key TEXT NOT NULL PRIMARY KEY, value TEXT NOT NULL)')
   dbPath = filepath
   console.log('Connected to the %s SQLite database.', filepath)
+  iterKeysStatement = db.prepareQuery<[string]>('SELECT key FROM Data')
+  readStatement = db.prepareQuery<[string]>('SELECT value FROM Data WHERE key = ?')
   // Use "upsert" syntax to store an entry only if the key is not already in the DB.
   // https://www.sqlite.org/lang_upsert.html
-  writeStatement = db.prepareQuery('INSERT INTO Data(key, value) VALUES(?, ?) ON CONFLICT (key) DO NOTHING')
+  writeOnceStatement = db.prepareQuery('INSERT INTO Data(key, value) VALUES(?, ?) ON CONFLICT (key) DO NOTHING')
+  writeStatement = db.prepareQuery('REPLACE INTO Data(key, value) VALUES(?, ?)')
+}
+
+export function count (): number {
+  return db.query<[number]>('SELECT COUNT(*) FROM Data')[0][0]
+}
+
+// deno-lint-ignore require-await
+export async function readData (key: string): Promise<Uint8Array | string | void> {
+  // For some reason `[null]` is returned when the value is an empty Uint8Array.
+  const maybeRow = readStatement.first([key])
+  return maybeRow === undefined ? undefined : maybeRow[0] ?? new Uint8Array()
 }
 
-export function writeData (key: string, value: Buffer | string): Promise<void> {
+export async function* iterKeys (): AsyncGenerator<string> {
+  for (const row of iterKeysStatement.iter()) {
+    yield row[0]
+  }
+}
+
+// deno-lint-ignore require-await
+export async function writeData (key: string, value: Uint8Array | string): Promise<void> {
   checkKey(key)
   writeStatement.execute([key, value])
 }
+
+// deno-lint-ignore require-await
+export async function writeDataOnce (key: string, value: Uint8Array | string): Promise<void> {
+  checkKey(key)
+  writeOnceStatement.execute([key, value])
+}
```

Comment:
The functions `readData`, `writeData`, and `writeDataOnce` are marked as `async` but don't call `await`.

If this because they are using a synchronous API and we want to ensure API compatibility across backends, can you add a comment to that effect? Without an explanation, it will look weird, especially if we remove the `fs` backend.

BTW, I think you can call `await` on a synchronous function, so they might not even need the `async` in front of them.

Reply:
Yea that's mainly for API compatibility across backends, and also as a visual clue that the function returns a Promise. Maybe I'm wrong, but I'm not fan of promise-returning functions not being explicitly marked as `async`.

> I think you can call await on a synchronous function
Yes, which means the same API function could either return a promise or a plain value according to different backends, and it would work as long as the caller doesn't forget to await the result. But I'm not sure the different backend modules would still have the same type signature.

Actually the `count()` function is not yet consistent. I don't know yet if it would be sync or async in most backends.

Reply:
> Actually the `count()` function is not yet consistent in that regard. I don't know yet if it would be sync or async in most backends.

I think we should assume on some backends it might be asynchronous